---
title: "Compare my Work and Yuan`s"
author: "Markus Heidrich"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# knitr::opts_chunk$set(echo = TRUE, fig.width = 12, fig.height = 12)
```

## Report
Compare and show details.

## Setup

### Load Libraries

libraries required to run this report.

```{r load-libraries}

### Bioconductor and CRAN libraries used
library(knitr)
library(ggplot2)
suppressMessages(library("CellTagR"))
suppressMessages(library("Seurat"))
suppressMessages(library("optparse"))
```

### Execution

```{r load-data}

markus = CellTagObject(object.name = "markus.obj", fastq.bam.directory = "data/bam/possorted_genome_bam.filtered_merged.bam")
markus = CellTagExtraction(markus, celltag.version = "v1")
markus =  CellTagMatrixCount(celltag.obj = markus, barcodes.file = "data/samples/SI-TT-H4/outs/filtered_feature_bc_matrix/barcodes.tsv")

yuan = CellTagObject(object.name = "yuan.obj", fastq.bam.directory = "Patrick_Anika_22-10/scRNAseq/07_extCellTag_Patrick_Anika/output/SI-TT-H4/outs/possorted_genome_bam.filtered.bam")
yuan = CellTagExtraction(yuan, celltag.version = "v1")
yuan = CellTagMatrixCount(celltag.obj = yuan, barcodes.file = "Patrick_Anika_22-10/scRNAseq/07_extCellTag_Patrick_Anika/output/SI-TT-H4/outs/filtered_feature_bc_matrix/barcodes.tsv")

print(markus)
print(yuan)

curr.mtx <- slot(markus, "raw.count")
curr.version <- markus@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[!(Matrix::rowSums(is.na(curr.mtx.sub)) == ncol(curr.mtx.sub) | Matrix::rowSums(curr.mtx.sub == 0) == ncol(curr.mtx.sub)),]
length(rownames(full.mtx.sub))

curr.mtx <- slot(yuan, "raw.count")
curr.version <- yuan@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[!(Matrix::rowSums(is.na(curr.mtx.sub)) == ncol(curr.mtx.sub) | Matrix::rowSums(curr.mtx.sub == 0) == ncol(curr.mtx.sub)),]
length(rownames(full.mtx.sub))

markus <- CellTagDataForCollapsing(celltag.obj = markus, output.file = "markus.txt")
yuan <- CellTagDataForCollapsing(celltag.obj = yuan, output.file = "yuan.txt")

saveRDS(markus, "markus.RDS")
saveRDS(markus, "yuan.RDS")

```

```{r, engine = 'bash', eval = FALSE}
starcode -s --print-clusters markus.txt > markus_result.txt
starcode -s --print-clusters yuan.txt > yuan_result.txt
```

```{r after_collapse}
markus = readRDS("markus.RDS")
yuan = readRDS("yuan.RDS")

markus <- CellTagDataPostCollapsing(celltag.obj = markus, collapsed.rslt.file = "markus_result.txt")
yuan <- CellTagDataPostCollapsing(celltag.obj = yuan, collapsed.rslt.file = "yuan_result.txt")

length(rownames(markus@collapsed.count))
length(rownames(yuan@collapsed.count))

curr.mtx <- slot(markus, "collapsed.count")
curr.version <- markus@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

curr.mtx <- slot(yuan, "collapsed.count")
curr.version <- yuan@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

markus <- SingleCellDataBinarization(markus, 1)
yuan <- SingleCellDataBinarization(yuan, 1)

curr.mtx <- slot(markus, "binary.mtx")
curr.version <- markus@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

curr.mtx <- slot(yuan, "binary.mtx")
curr.version <- yuan@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

dt.mtx.whitelist.path <- system.file("extdata", "v1_whitelist.csv", package = "CellTagR")
markus <- SingleCellDataWhitelist(markus, dt.mtx.whitelist.path)
yuan <- SingleCellDataWhitelist(yuan, dt.mtx.whitelist.path)

curr.mtx <- slot(markus, "whitelisted.count")
curr.version <- markus@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

curr.mtx <- slot(yuan, "whitelisted.count")
curr.version <- yuan@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

curr.mtx <- slot(markus, "whitelisted.count")
curr.version <- markus@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[!(Matrix::rowSums(is.na(curr.mtx.sub)) == ncol(curr.mtx.sub) | Matrix::rowSums(curr.mtx.sub == 0) == ncol(curr.mtx.sub)),]
length(rownames(full.mtx.sub))

curr.mtx <- slot(yuan, "whitelisted.count")
curr.version <- yuan@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[!(Matrix::rowSums(is.na(curr.mtx.sub)) == ncol(curr.mtx.sub) | Matrix::rowSums(curr.mtx.sub == 0) == ncol(curr.mtx.sub)),]
length(rownames(full.mtx.sub))


markus@metric.filtered.count <- as(matrix(NA, 0, 0), "dgCMatrix")
markus <- MetricBasedFiltering(markus, 20, comparison = "less")
yuan@metric.filtered.count <- as(matrix(NA, 0, 0), "dgCMatrix")
yuan <- MetricBasedFiltering(yuan, 20, comparison = "less")

curr.mtx <- slot(markus, "metric.filtered.count")
curr.version <- markus@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

curr.mtx <- slot(yuan, "metric.filtered.count")
curr.version <- yuan@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

# Filter out cells with less than 2 CellTags
markus <- MetricBasedFiltering(markus, 1, comparison = "greater")
yuan <- MetricBasedFiltering(yuan, 1, comparison = "greater")

curr.mtx <- slot(markus, "metric.filtered.count")
curr.version <- markus@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

curr.mtx <- slot(yuan, "metric.filtered.count")
curr.version <- yuan@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

markus <- JaccardAnalysis(markus)
yuan <- JaccardAnalysis(yuan)

curr.mtx <- slot(markus, "jaccard.mtx")
curr.version <- markus@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

curr.mtx <- slot(yuan, "jaccard.mtx")
curr.version <- yuan@curr.version
curr.mtx.sub <- curr.mtx[, which(startsWith(colnames(curr.mtx), curr.version))]
colnames(curr.mtx.sub) <- gsub(pattern = paste0(curr.version, "."), replacement = "", colnames(curr.mtx.sub))
full.mtx.sub <- curr.mtx.sub[Matrix::rowSums(is.na(curr.mtx.sub)) != ncol(curr.mtx.sub),]
length(rownames(full.mtx.sub))

markus <- CloneCalling(celltag.obj = markus, correlation.cutoff=0.7)
yuan <- CloneCalling(celltag.obj = yuan, correlation.cutoff=0.7)

length(rownames(markus@clone.composition$v1$cell.barcode))
length(rownames(yuan@clone.composition$v1$cell.barcode))

```

Jaccard analysis using the R package Proxy was used to calculate the similarity of CellTag signatures between cells. A Jaccard score of >0.7 was used as a cut-off to identify cells highly likely to be related, on the basis of our experimental findings.
